framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -381,9 +379,55 @@ public void processClassTree(ClassTree classTree) {
 
         checkExtendsImplements(classTree);
 
+        checkPolymorphicClass(classTree);
+
         super.visitClass(classTree, null);
     }
 
+    /**
+     * Throws an error if {@code classTree} has polymorphic fields but is not annotated with
+     * {@code @HasQualifierParameter}.
+     */
+    private void checkPolymorphicClass(ClassTree classTree) {
+        boolean hasPolyField = false;
+        List<? extends Tree> members = classTree.getMembers();
+        Set<? extends AnnotationMirror> topAnnotations =
+                atypeFactory.getQualifierHierarchy().getTopAnnotations();
+
+        for (Tree mem : members) {
+            if (mem.getKind() == Tree.Kind.VARIABLE) {
+                AnnotatedTypeMirror fieldAnno = atypeFactory.getAnnotatedType(mem);
+                for (AnnotationMirror top : topAnnotations) {
+                    AnnotationMirror polyAnnoInHierarchy =
+                            atypeFactory.getQualifierHierarchy().getPolymorphicAnnotation(top);
+                    if (polyAnnoInHierarchy != null) {
+                        AnnotationMirror fieldAnnoInHierarchy =
+                                fieldAnno.getAnnotationInHierarchy(top);
```

Comment:
This is only looking at the primary annotation. An error would not be issued if the field was @Poly String[].

Reply:
You can use an AnnotatedTypeScanner To do this.

Reply:
@smillst Is there an easy way of getting the enclosing class element from inside AnnotatedTypeScanner?

Reply:
The enclosing class of what?

Reply:
org.checkerframework.common.basetype.BaseTypeVisitor#checkContainsSameToString is a good example of using an AnnotatedTypeScanner (or in this case a SimpleAnnotatedTypeScanner).

Reply:
> The enclosing class of what?

If a field is annotated `@Poly`, we want to check that the enclosing class is annotated with `@HasQualifierParameter`. I thought you meant I should be able to do this check in AnnotatedTypeScanner. Am I missing something?

Reply:
I mean you can use an AnnotatedTypeScanner, or probably a SimpleAnnotatedTypeScanner, to check if the type of the field has a poly annotation.  If so, then check if the enclosing class has `@HasQualifierParameter`.  Then issue an error for the field.

Reply:
Git it. Thanks!